flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -762,17 +762,21 @@ private void doRun() {
             executingThread.setContextClassLoader(userCodeClassLoader.asClassLoader());
 
             AbstractInvokable finalInvokable = invokable;
-            runWithSystemExitMonitoring(finalInvokable::restore);
+            try {
+                runWithSystemExitMonitoring(finalInvokable::restore);
 
-            if (!transitionState(ExecutionState.INITIALIZING, ExecutionState.RUNNING)) {
-                throw new CancelTaskException();
-            }
+                if (!transitionState(ExecutionState.INITIALIZING, ExecutionState.RUNNING)) {
+                    throw new CancelTaskException();
+                }
 
-            // notify everyone that we switched to running
-            taskManagerActions.updateTaskExecutionState(
-                    new TaskExecutionState(executionId, ExecutionState.RUNNING));
+                // notify everyone that we switched to running
+                taskManagerActions.updateTaskExecutionState(
+                        new TaskExecutionState(executionId, ExecutionState.RUNNING));
 
-            runWithSystemExitMonitoring(finalInvokable::invoke);
+                runWithSystemExitMonitoring(finalInvokable::invoke);
+            } finally {
+                runWithSystemExitMonitoring(finalInvokable::cleanUp);
+            }
```

Comment:
Thanks for sharing and raising this!
I like the idea and I agree we should improve the contract `AbstractInvokable` and not postpone it till larger refactorings.

Maybe we should name it just `Invokable` or `TaskInvokable` (as it's actually invokable part of the `Task`)?
`cancel()` I think should also belong to this interface.

Another set of methods I'd extract are related to checkpointing (those throwing `UnsupportedOperationException` currently). We could have `CheckpointableInvokable extends AbstractInvokable` implemented only by a subset of classes.

Reply:
Super, it will be great if you will do that. Both names `Invokable` and `TaskInvokable` looks good to me but perhaps `TaskInvokable` is a little better since it is more specific and give more information about usage which is good.

Reply:
I've extracted the interfaces in 9c217799fb4aaeb5bee3fcb1f1408f3842c66b54
and used them more widely in 7665ff55b614b610c6a41a708e7e9d4a8c736fa7.

WDYT of this change?

The commits change about 20 files but I'm still not sure if it should be included into 1.14.

cc: @pnowojski

Reply:
I would merge this to both branches. It doesn't seems as a very risky refactor. But that's a question to the release managers.

Reply:
For the record: 
I dicussed it with @xintongsong offline. There were no concerns regarding merging the bugfix as well as the refactoring if it's necessary.